File: valo_tui/screens/match_detail.py
"""Match detail v2 — series header, map vetoes, and per-map sections with
win-bars, round-momentum sparklines, agent-role icons, and skeleton rows for
maps that haven't started.

The series fetch can hit the network (read-through cache), so it runs in a
worker thread to keep the UI responsive."""

from __future__ import annotations

from rich.text import Text

from textual import work
from textual.app import ComposeResult
from textual.binding import Binding
from textual.containers import Horizontal, VerticalScroll
from textual.screen import Screen
from textual.widgets import Footer, Label

from ..data import cache
from ..data.models import MapScore, SeriesDetail
from ..style import bars
from ..style.icons import agent_glyph, map_icon
from .widgets import LIVE, MUTED, TEXT, LiveDot, SkeletonRow, VimDataTable

_COLS = [
    ("player", 16),
    ("agent", 13),
    ("acs", 5),
    ("k", 4),
    ("d", 4),
    ("a", 4),
    ("adr", 5),
    ("hs%", 5),
    ("fk", 4),
    ("fd", 4),
]


class MatchDetailScreen(Screen):
    BINDINGS = [
        Binding("escape,q", "app.pop_screen", "Back"),
        Binding("j", "scroll_down", "Down", show=False),
        Binding("k", "scroll_up", "Up", show=False),
    ]

    def __init__(self, match_id: int) -> None:
        super().__init__()
        self.match_id = match_id

    def compose(self) -> ComposeResult:
        with VerticalScroll(id="detail-scroll"):
            yield SkeletonRow(cells=6)
            yield Label(f"[{MUTED}]loading match {self.match_id}…[/]")
        yield Footer()

    def on_mount(self) -> None:
        self._load()

    @work(thread=True, exclusive=True)
    def _load(self) -> None:
        detail = cache.series_detail(self.match_id)
        self.app.call_from_thread(self._render_detail, detail)

    def _render_detail(self, detail: SeriesDetail | None) -> None:
        body = self.query_one("#detail-scroll", VerticalScroll)
        body.remove_children()
        if detail is None:
            body.mount(Label(f"[{MUTED}]no detail available for this match[/]"))
            return

        self._mount_header(body, detail)
        if detail.vetoes:
            body.mount(Label(self._vetoes(detail), classes="veto"))

        maps = [m for m in detail.maps if not m.is_aggregate]
        if not maps:
            body.mount(Label(f"[{MUTED}]no map data yet[/]"))
            return

        for m in maps:
            self._mount_map(body, detail, m)

    # ── header ───────────────────────────────────────────────
    def _mount_header(self, body: VerticalScroll, d: SeriesDetail) -> None:
        s1 = d.team1.score if d.team1.score is not None else "–"
        s2 = d.team2.score if d.team2.score is not None else "–"
        title = (
            f"[bold {TEXT}]{d.team1.name}[/]  [{LIVE}]{s1} – {s2}[/]  "
            f"[bold {TEXT}]{d.team2.name}[/]"
        )
        bo = f"  ·  [{MUTED}]{d.best_of}[/]" if d.best_of else ""
        if d.is_live:
            status_widget = LiveDot("LIVE")
        elif d.is_completed:
            status_widget = Label(f"[{MUTED}]✓ final[/]")
        else:
            status = d.remaining or d.status_note or "upcoming"
            status_widget = Label(f"[{LIVE}]○ {status}[/]")
        body.mount(
            Horizontal(
                status_widget,
                Label(f"  {title}{bo}"),
                classes="series-header",
            )
        )
        body.mount(Label(f"[{MUTED}]{d.event} · {d.phase}[/]", classes="detail-sub"))

    def _vetoes(self, d: SeriesDetail) -> str:
        parts = []
        for v in d.vetoes:
            verb = v.action.lower()
            icon = map_icon(v.map)
            if verb == "ban":
                parts.append(f"[{MUTED}]✖ {v.team} ban {icon} {v.map}[/]")
            elif verb == "pick":
                parts.append(f"[{TEXT}]➤ {v.team} pick {icon} {v.map}[/]")
            else:
                parts.append(f"[{MUTED}]➤ {icon} {v.map} (decider)[/]")
        return "veto:  " + "   ".join(parts)

    # ── per-map section ──────────────────────────────────────
    def _mount_map(self, body: VerticalScroll, d: SeriesDetail, m: MapScore) -> None:
        body.mount(Label(self._map_header(d, m), classes="map-title"))
        if m.state == "pending":
            for _ in range(3):
                body.mount(SkeletonRow(cells=5))
            return
        if m.rounds:
            body.mount(
                Label(f"  {bars.momentum(m.rounds, m.team1_short)}", classes="momentum")
            )
        body.mount(self._map_table(m))

    def _map_header(self, d: SeriesDetail, m: MapScore) -> str:
        icon = map_icon(m.name)
        bar = bars.winbar(m.team1_score, m.team2_score)
        if m.has_score:
            score = f"[{TEXT}]{m.team1_score}–{m.team2_score}[/]"
        else:
            score = f"[{MUTED}]TBD[/]"
        pick = d.pick_label(m.name)
        pick_txt = f"   [{MUTED}]{pick}[/]" if pick else ""
        return f"{icon} [{TEXT}]{m.name:<9}[/] {bar}  {score}{pick_txt}"

    def _map_table(self, m: MapScore) -> VimDataTable:
        table = VimDataTable(cursor_type="row", zebra_stripes=False)
        for name, width in _COLS:
            table.add_column(name, width=width)
        for p in sorted(m.players, key=lambda p: (p.acs or 0), reverse=True):
            table.add_row(
                Text(p.name, style=TEXT),
                self._agent_cell(p.agents),
                _num(p.acs),
                _num(p.k),
                _num(p.d),
                _num(p.a),
                _num(p.adr),
                _pct(p.hs_pct),
                _num(p.fk),
                _num(p.fd),
            )
        table.styles.height = len(m.players) + 1
        return table

    def _agent_cell(self, agents: list[str]) -> Text:
        if not agents:
            return Text("—", style=MUTED)
        glyph, colour = agent_glyph(agents[0])
        cell = Text()
        cell.append(f"{glyph} ", style=colour)
        cell.append(", ".join(agents), style=MUTED)
        return cell


def _num(v) -> Text:
    return Text("–" if v is None else (f"{v:.0f}" if isinstance(v, float) else str(v)))


def _pct(v) -> Text:
    return Text("–" if v is None else f"{v:.0f}%")

File: internal/screens/match_detail.go
package screens

import (
	"fmt"
	"image/color"
	"sort"
	"strings"

	tea "charm.land/bubbletea/v2"
	"charm.land/lipgloss/v2"

	"github.com/jashkarangiya/valo-tui/internal/data"
	"github.com/jashkarangiya/valo-tui/internal/styles"
)

var (
	team1Col = styles.Team1 // left team (Valorant red)
	team2Col = styles.Team2 // right team (teal)
)

var bracketPhases = []string{
	"playoff", "bracket", "upper", "lower", "grand final",
	"quarterfinal", "semifinal", "final",
}
var roleOrder = []string{"duelist", "initiator", "controller", "sentinel", "flex"}

// MatchDetail is the tactical broadcast overlay: hero score, series momentum,
// per-map scoreboards. Mirrors screens/match_detail.py.
type MatchDetail struct {
	matchID int
	w, h    int
	scroll  int
	detail  data.SeriesDetail
	ok      bool
}

func NewMatchDetail(matchID, w, h int) MatchDetail {
	m := MatchDetail{matchID: matchID, w: w, h: h}
	m.detail, m.ok = data.SeriesDetailFor(matchID)
	return m
}

func (m *MatchDetail) SetSize(w, h int) { m.w, m.h = w, h }

// TeamAt returns the team name for a click at overlay-local (x, y), so the
// header names open a roster. Only the unscrolled header band is hot; the
// left/right half of the width picks the team (names sit either side of "vs").
func (m MatchDetail) TeamAt(x, y int) (string, bool) {
	if !m.ok || m.scroll != 0 || y > 5 {
		return "", false
	}
	if x < m.w/2 {
		return m.detail.Team1.Name, true
	}
	return m.detail.Team2.Name, true
}

func (m MatchDetail) hasBracket() bool {
	if !m.ok {
		return false
	}
	p := strings.ToLower(m.detail.Phase)
	for _, k := range bracketPhases {
		if strings.Contains(p, k) {
			return true
		}
	}
	return false
}

func (m MatchDetail) Update(msg tea.Msg) (MatchDetail, tea.Cmd) {
	if k, ok := msg.(tea.KeyPressMsg); ok {
		switch k.String() {
		case "esc", "q":
			return m, func() tea.Msg { return CloseOverlayMsg{} }
		case "b":
			if m.hasBracket() {
				name := m.detail.Event
				return m, func() tea.Msg { return OpenBracketMsg{EventName: name} }
			}
		case "j", "down":
			m.scroll++
		case "k", "up":
			if m.scroll > 0 {
				m.scroll--
			}
		}
	}
	return m, nil
}

func (m MatchDetail) View() string {
	body := m.render()
	lines := strings.Split(body, "\n")
	// Reserve one row for the footer hint.
	visible := m.h - 1
	if visible < 1 {
		visible = 1
	}
	if m.scroll > len(lines)-visible {
		m.scroll = max(0, len(lines)-visible)
	}
	end := min(len(lines), m.scroll+visible)
	view := strings.Join(lines[m.scroll:end], "\n")
	footer := mutedSt.Render("j/k scroll · click team → roster · esc back")
	if m.hasBracket() {
		footer = mutedSt.Render("j/k scroll · b bracket · click team → roster · esc back")
	}
	return view + "\n" + footer
}

func (m MatchDetail) render() string {
	if !m.ok {
		return muted(fmt.Sprintf("no detail available for match %d", m.matchID))
	}
	d := m.detail
	var b strings.Builder
	b.WriteString(m.header(d))

	var maps []data.MapScore
	for _, mp := range d.Maps {
		if !mp.IsAggregate() {
			maps = append(maps, mp)
		}
	}
	b.WriteString(m.seriesMomentum(d, maps))
	if len(maps) == 0 {
		b.WriteString("\n" + muted("no map data yet"))
		return b.String()
	}
	for _, mp := range maps {
		b.WriteString(m.mapBlock(d, mp))
	}
	return b.String()
}

func (m MatchDetail) header(d data.SeriesDetail) string {
	s1, s2 := derefOr0(d.Team1.Score), derefOr0(d.Team2.Score)
	var status string
	switch {
	case d.IsLive():
		status = lipgloss.NewStyle().Foreground(team1Col).Render("● live")
	case d.IsCompleted():
		status = muted("✓ final")
	default:
		r := d.Remaining
		if r == "" {
			r = "upcoming"
		}
		status = muted("○ " + r)
	}

	names := lipgloss.NewStyle().Foreground(team1Col).Bold(true).Render(d.Team1.Name) +
		"    " + muted("vs") + "    " +
		lipgloss.NewStyle().Foreground(team2Col).Bold(true).Render(d.Team2.Name)

	left := lipgloss.NewStyle().Foreground(team1Col).Bold(true).Render(big(fmt.Sprint(s1)))
	right := lipgloss.NewStyle().Foreground(team2Col).Bold(true).Render(big(fmt.Sprint(s2)))
	dash := lipgloss.NewStyle().Foreground(styles.Muted).Render("\n\n  —  ")
	hero := lipgloss.JoinHorizontal(lipgloss.Center, left, dash, right)

	meta := ""
	if d.BestOf != "" {
		meta = muted(d.BestOf)
	}
	if d.Phase != "" {
		if meta != "" {
			meta += muted(" · " + d.Phase)
		} else {
			meta = muted(d.Phase)
		}
	}
	if meta != "" {
		meta = meta + "   " + status
	} else {
		meta = status
	}
	if m.hasBracket() {
		meta += "    " + lipgloss.NewStyle().Foreground(team1Col).Render("[b]") + " " + text("bracket")
	}

	center := func(s string) string { return lipgloss.PlaceHorizontal(m.w, lipgloss.Center, s) }
	var b strings.Builder
	b.WriteString(center(names) + "\n")
	b.WriteString(center(hero) + "\n")
	b.WriteString(center(meta) + "\n")
	b.WriteString(center(muted(d.Event)) + "\n")
	if intel := m.intel(d); intel != "" {
		b.WriteString(intel + "\n")
	}
	b.WriteString("\n")
	return b.String()
}

func (m MatchDetail) intel(d data.SeriesDetail) string {
	var players []data.PlayerLine
	for _, mp := range d.Maps {
		players = append(players, mp.Players...)
	}
	if len(players) == 0 {
		return ""
	}
	top, fk := players[0], players[0]
	for _, p := range players {
		if derefOr0(p.ACS) > derefOr0(top.ACS) {
			top = p
		}
		if derefOr0(p.FK) > derefOr0(fk.FK) {
			fk = p
		}
	}
	return muted("intel:  top acs ") + text(fmt.Sprintf("%s %d", top.Name, derefOr0(top.ACS))) +
		muted("   ·   most FK ") + text(fmt.Sprintf("%s %d", fk.Name, derefOr0(fk.FK)))
}

func (m MatchDetail) seriesMomentum(d data.SeriesDetail, maps []data.MapScore) string {
	var played []data.MapScore
	for _, mp := range maps {
		if mp.HasScore() {
			played = append(played, mp)
		}
	}
	if len(played) == 0 {
		return ""
	}
	var b strings.Builder
	b.WriteString(lipgloss.NewStyle().Foreground(team1Col).Render("series momentum") + "\n")
	for _, mp := range played {
		bar := styles.WinBar(mp.Team1Score, mp.Team2Score, 18)
		pick := d.PickLabel(mp.Name)
		pickTxt := ""
		if pick != "" {
			pickTxt = "   " + muted(pick)
		}
		line := styles.MapIcon(mp.Name) + " " + text(fmt.Sprintf("%-9s", mp.Name)) + " " + bar + "  " +
			text(fmt.Sprintf("%d–%d", derefOr0(mp.Team1Score), derefOr0(mp.Team2Score))) + pickTxt
		b.WriteString(line + "\n")
	}
	b.WriteString("\n")
	return b.String()
}

func (m MatchDetail) mapBlock(d data.SeriesDetail, mp data.MapScore) string {
	var b strings.Builder
	b.WriteString(m.mapTitle(d, mp) + "\n")
	if mp.State() == "pending" {
		b.WriteString(muted("  (not played yet)") + "\n\n")
		return b.String()
	}
	if len(mp.Rounds) > 0 {
		b.WriteString("  rounds  " + styles.Momentum(mp.Rounds, mp.Team1Short) + "\n")
	}
	b.WriteString(m.scoreboard(mp))
	return b.String()
}

func (m MatchDetail) mapTitle(d data.SeriesDetail, mp data.MapScore) string {
	score := muted("TBD")
	if mp.HasScore() {
		score = text(fmt.Sprintf("%d–%d", derefOr0(mp.Team1Score), derefOr0(mp.Team2Score)))
	}
	pick := d.PickLabel(mp.Name)
	pickTxt := ""
	if pick != "" {
		pickTxt = "   " + muted(pick)
	}
	return "\n" + styles.MapIcon(mp.Name) + " " + textB(mp.Name) + "   " + score + pickTxt
}

func (m MatchDetail) scoreboard(mp data.MapScore) string {
	teams := []struct {
		short  string
		colour color.Color
	}{{mp.Team1Short, team1Col}, {mp.Team2Short, team2Col}}

	hdr := mutedSt.Render(fmt.Sprintf("%-15s%4s %3s %3s %3s %5s %4s", "", "acs", "k", "d", "a", "adr", "hs"))
	var b strings.Builder
	for _, t := range teams {
		var tp []data.PlayerLine
		for _, p := range mp.Players {
			if p.TeamShort == t.short {
				tp = append(tp, p)
			}
		}
		if len(tp) == 0 {
			continue
		}
		short := t.short
		if short == "" {
			short = "?"
		}
		b.WriteString(lipgloss.NewStyle().Foreground(t.colour).Bold(true).Render(short) + "\n")
		b.WriteString(hdr + "\n")

		byRole := map[string][]data.PlayerLine{}
		for _, p := range tp {
			role := styles.AgentRole(p.Agent())
			if role == "" {
				role = "flex"
			}
			byRole[role] = append(byRole[role], p)
		}
		for _, role := range roleOrder {
			ps := byRole[role]
			if len(ps) == 0 {
				continue
			}
			b.WriteString("  " + muted(role+"s") + "\n")
			sort.SliceStable(ps, func(a, bb int) bool { return derefOr0(ps[a].ACS) > derefOr0(ps[bb].ACS) })
			for _, p := range ps {
				b.WriteString(m.playerLine(p, t.colour) + "\n")
			}
		}
		b.WriteString("\n")
	}
	return b.String()
}

func (m MatchDetail) playerLine(p data.PlayerLine, colour color.Color) string {
	glyph, gcol := styles.AgentGlyph(p.Agent())
	agent := p.Agent()
	if agent == "" {
		agent = "—"
	}
	agent = clipRunes(agent, 8)
	name := clipRunes(p.Name, 12)
	nameCol := lipgloss.NewStyle().Foreground(colour).Render(fmt.Sprintf("%-12s", name))
	glyphCol := lipgloss.NewStyle().Foreground(gcol).Render(glyph)
	stats := fmt.Sprintf(" %4s %3s %3s %3s %5s %4s", n(p.ACS), n(p.K), n(p.D), n(p.A), fNum(p.ADR), pct(p.HSPct))
	return "    " + nameCol + " " + glyphCol + " " + mutedSt.Render(fmt.Sprintf("%-8s", agent)) + textSt.Render(stats)
}

// ── numeric formatting (mirror _n / _f / _pct) ──────────────

func n(v *int) string {
	if v == nil {
		return "–"
	}
	return fmt.Sprint(*v)
}

func fNum(v *float64) string {
	if v == nil {
		return "–"
	}
	return fmt.Sprintf("%.0f", *v)
}

func pct(v *float64) string {
	if v == nil {
		return "–"
	}
	return fmt.Sprintf("%.0f%%", *v)
}

func derefOr0(p *int) int {
	if p == nil {
		return 0
	}
	return *p
}

func clipRunes(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
